### DISCUSSION TITLE
Compiling Application on HPC

### DISCUSSION THREAD

**Discussion by jazorn** [idaholab/moose#25092] [Q&A Getting Started]
Looking for some help getting an application up and running using MOOSE. 

My HPC Administrator has installed MOOSE following the HPC Cluster instructions (https://mooseframework.inl.gov/getting_started/installation/hpc_install_moose.html) and installed it as a module that can be loaded (like typical HPC Modules, i.e. module load moose).

However, when I go to compile/install a new application (in particular, MACAW [https://github.com/marinasessim/macaw/tree/main ]), I get compilation errors because it tries to create directories in HPC locations that I can't create directories in. My initial thought was that I needed to catch the location of the moose directory that these makefiles access but that didn't seem to fix the problem.

I am currently away from my work computer, but when I get back with access to my HPC I will post some examples of the errors I am receiving.

Any help would be greatly appreciated. 

UPDATE:

Here is an example of an error I am getting when running the Makefile of the MACAW Application. 

UPDATE 2:

Noticed in some other discussions that images of errors are frowned upon. See the text version of the error below

`Creating Unity Directory /opt/Software/mooseproject/moose/modules/ray_tracing/build/unity_src
mkdir: cannot create directory ‘/opt/Software/mooseproject/moose/modules/ray_tracing/build’: Permission denied
make: *** [/opt/Software/mooseproject/moose/modules/ray_tracing/build/unity_src] Error 1`

Also I saw in the main moose directory (located at `/opt/Software/mooseproject/moose/` on my HPC system), there was a configure script. So I went back and tried running that. Both in the moose directory and in my application directory and still get errors. 

Apologies if it seems like I am trying to spam this channel, trying to keep the discussion up-to-date with different things I have tried so in the future perhaps this helps others.

Cheers.

**Comment by milljm** [comment]
Those instructions are meant to be followed by you, and not the system admin.

As you are running into, you need to build your own copy of MOOSE, along side your application. The system administrator should only be required to provide you a means of compiling your code (the MPI stack). They should not be responsible for supplying you pre-built versions of MOOSE.

To do so would require that they also:

- Allow all users of MOOSE to have write permissions to the source directory (bad... very bad idea)
- Update MOOSE whenever:
  - There is an update to MOOSE and other users on your system need it (multiple times a day!)
  - There is a libMesh update (twice a month)
  - There is a PETSc update (twice a year-ish)

...which would _force_ all users of MOOSE on this system to also rebuild their applications.

Basically, it is not feasible to provide a one-version-for-everyone MOOSE to multiple users on a system.

What you need to figure out is how the system admin built MOOSE. You can achieve this by loading the `moose` environment, and then seeing what else gets loaded, and then loading everything except moose.

Example:

```bash
module purge    # clean up any modules already being loaded
module load moose
module list

$ module list

Currently Loaded Modules:
  1) use.moose                    5) patchelf/0.9-8.5.0
  2) moose-apps                   6) git/2.25.0-gcc-9.2.0-v4mv
  3) gcc/8.4.0-gcc-4.8.5-jacd     7) miniconda
  4) mvapich2/2.3.7-1-gcc-8.4.0   8) moose/2023-07-31
```
This tells me that I need to actually load modules "1-7" in order to be in an environment where I can build MOOSE.
